homeassistant/components/file_upload/__init__.py

```diff
@@ -0,0 +1,176 @@
+"""The File Upload integration."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Iterator
+from contextlib import contextmanager
+from dataclasses import dataclass
+from pathlib import Path
+import shutil
+import tempfile
+
+from aiohttp import web
+import voluptuous as vol
+
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.components.http.data_validator import RequestDataValidator
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util.ulid import ulid_hex
+
+DOMAIN = "file_upload"
+
+# If increased, change upload view to streaming
+# https://docs.aiohttp.org/en/stable/web_quickstart.html#file-uploads
+MAX_SIZE = 1024 * 1024 * 10
+TEMP_DIR_NAME = f"home-assistant-{DOMAIN}"
+
+
+@contextmanager
+def process_uploaded_file(hass: HomeAssistant, file_id: str) -> Iterator[Path]:
+    """Get an uploaded file.
+
+    File is removed at the end of the context.
+    """
+    if DOMAIN not in hass.data:
+        raise ValueError("File does not exist")
+
+    file_upload_data: FileUploadData = hass.data[DOMAIN]
+
+    if not file_upload_data.has_file(file_id):
+        raise ValueError("File does not exist")
+
+    try:
+        yield file_upload_data.file_path(file_id)
+    finally:
+        file_upload_data.files.pop(file_id)
+        shutil.rmtree(file_upload_data.file_dir(file_id))
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up File Upload."""
+    hass.http.register_view(FileUploadView)
+    return True
+
+
+@dataclass(frozen=True)
+class FileUploadData:
+    """File upload data."""
+
+    temp_dir: Path
+    files: dict[str, str]
+
+    @classmethod
+    async def create(cls, hass: HomeAssistant) -> FileUploadData:
+        """Initialize the file upload data."""
+
+        def _create_temp_dir() -> Path:
+            """Create temporary directory."""
+            temp_dir = Path(tempfile.gettempdir()) / TEMP_DIR_NAME
+
+            # If it exists, it's an old one and Home Assistant didn't shut down correctly.
+            if temp_dir.exists():
+                shutil.rmtree(temp_dir)
+
+            temp_dir.mkdir()
```

Comment:
Since we don't support Windows, `tempfile.mkdtemp` is more or less this:

```suggestion
            temp_dir.mkdir(0o700)
```

Alternatively, could we directly use `tempfile.TemporaryDirectory` and call its `cleanup` method?

Reply:
I was under the impression `tempfile.TemporaryDirectory()` does the same as what I do here?

Reply:
Ah I see now that `mkdtemp` adds the `0o700`. Adding that.

The reason I use `gettempdir` directly is that I want to re-use, and cleanup old ones if necessary, the temporary directory. If HA plug is pulled, I would want the old files to be cleaned up.